usr/lib/common/pkey_utils.c

```diff
@@ -0,0 +1,531 @@
+{
+    CK_RV ret;
+
+    /* Unlock obj to remove READ_LOCK */
+    ret = object_unlock(key_obj);
+    if (ret != CKR_OK) {
+        TRACE_ERROR("object_unlock failed with rc=0x%lx\n", ret);
+        return ret;
+    }
+
+    /* Obtain write lock on key object */
+    ret = object_lock(key_obj, WRITE_LOCK);
+    if (ret != CKR_OK) {
+        TRACE_ERROR("object_lock for WRITE failed with rc=0x%lx\n", ret);
+        /* Try to restore read lock without rc checking */
+        object_lock(key_obj, READ_LOCK);
+        return ret;
+    }
+
+    /* Update attribute */
+    ret = template_update_attribute(key_obj->template, pkey_attr);
+    if (ret != CKR_OK) {
+        TRACE_ERROR("template_update_attribute failed with rc=0x%lx\n", ret);
+        goto done;
+    }
+
+    /* Save to repository if it's a token object */
+    if (object_is_token_object(key_obj)) {
+        ret = object_mgr_save_token_object(tokdata, key_obj);
+        if (ret != CKR_OK) {
+            TRACE_ERROR("Could not save token obj to repository, rc=0x%lx.\n", ret);
+        }
+    }
+
+done:
+
+    /* Restore object's READ_LOCK */
+    ret = object_unlock(key_obj);
+    if (ret != CKR_OK) {
+        TRACE_ERROR("object_unlock failed with rc=0x%lx\n", ret);
+    } else {
+        ret = object_lock(key_obj, READ_LOCK);
+        if (ret != CKR_OK)
+            TRACE_ERROR("object_lock for READ failed with rc=0x%lx\n", ret);
+    }
+
+    return ret;
+}
+
+/**
+ * Returns true if the protected key operation implied by the given mechanism
+ * is supported by CPACF, false otherwise.
+ */
+CK_BBOOL pkey_op_supported_by_cpacf(CK_MECHANISM *mech)
+{
+    if (!mech)
+        return CK_FALSE;
+
+    switch (mech->mechanism) {
+    case CKM_AES_ECB:
+    case CKM_AES_CBC:
+    case CKM_AES_CBC_PAD:
+    case CKM_AES_CMAC_GENERAL:
+    case CKM_AES_CMAC:
+        return CK_TRUE;
+    default:
+        return CK_FALSE;
+    }
+}
+
+/**
+ * Return the KM/KMC/KMAC function code for the given key length and
+ * encrypt/decrypt op.
+ */
+unsigned long get_function_code(CK_ULONG clear_keylen, CK_BYTE encrypt)
+{
+    unsigned long fc;
+
+    switch (clear_keylen) {
+    case 16:
+        fc = Encrypted_AES_128;
+        break;
+    case 24:
+        fc = Encrypted_AES_192;
+        break;
+    case 32:
+        fc = Encrypted_AES_256;
+        break;
+    default:
+        return 0;
+    }
+
+    if (!encrypt)
+        fc |= CPACF_DECRYPT;
+
+    return fc;
+}
+
+/**
+ * Performs an aes-ecb operation on the given data, using a protected key
+ * via the KM-encrypted-AES instruction. The protected key must be
+ * available in the key template as CKA_IBM_OPAQUE_PKEY.
+ */
+CK_RV pkey_aes_ecb(OBJECT *key_obj, CK_BYTE *in_data,
+                   CK_ULONG in_data_len, CK_BYTE *out_data,
+                   CK_ULONG_PTR p_output_data_len, CK_BYTE encrypt)
+{
+    CK_RV ret;
+    unsigned long fc;
+    CK_ATTRIBUTE *pkey_attr = NULL;
+    CK_ULONG clear_keylen;
+    struct __attribute__((packed)){
```

Comment:
Why packed?

Reply:
Why a struct at all?

Reply:
Well, it gets clearer when looking at the pkey_aes_cbc function below, where we have an additional iv field. The CPACF param block is different for the various functions, but for me it makes sense to code it in a very similar way. As the parm block goes to CPACF, I would always make it packed.